pkg/internalapis/csinodetopology/csinodetopology.go

```diff
@@ -16,9 +16,240 @@ limitations under the License.
+	volumeTopologyInstance *volumeTopology
+	// volumeTopologyInstanceLock is used for handling race conditions during read, write on volumeTopologyInstance
+	volumeTopologyInstanceLock = &sync.RWMutex{}
 )
+
+// InitTopologyServiceInterface returns a singleton implementation of the
+// TopologyService interface.
+func InitTopologyServiceInterface(ctx context.Context) (TopologyService, error) {
+	log := logger.GetLogger(ctx)
+
+	volumeTopologyInstanceLock.RLock()
+	if volumeTopologyInstance == nil {
+		volumeTopologyInstanceLock.RUnlock()
+		volumeTopologyInstanceLock.Lock()
+		defer volumeTopologyInstanceLock.Unlock()
+		if volumeTopologyInstance == nil {
+
+			// Get in cluster config for client to API server.
+			config, err := k8s.GetKubeConfig(ctx)
+			if err != nil {
+				log.Errorf("failed to get kubeconfig with error: %v", err)
+				return nil, err
+			}
+
+			// Create client to API server.
+			k8sclient, err := k8s.NewClientForGroup(ctx, config, csinodetopologyv1alpha1.GroupName)
+			if err != nil {
+				log.Errorf("failed to create k8sClient with error: %v", err)
+				return nil, err
+			}
+
+			// Create watcher for CSINodeTopology instances
+			crWatcher, err := k8s.NewCSINodeTopologyWatcher(ctx, config)
+			if err != nil {
+				log.Errorf("failed to create a watcher for CSINodeTopology CR. Error: %+v", err)
+				return nil, err
+			}
+
+			volumeTopologyInstance = &volumeTopology{
+				k8sClient:              k8sclient,
+				csiNodeTopologyWatcher: crWatcher,
+			}
+		}
+	} else {
+		volumeTopologyInstanceLock.RUnlock()
+	}
+
+	return volumeTopologyInstance, nil
+}
+
+// GetNodeTopologyLabels uses the CSINodeTopology CR to retrieve topology information of a node.
+func (volTopology *volumeTopology) GetNodeTopologyLabels(ctx context.Context, info *NodeInfo) (
+	map[string]string, error) {
+	log := logger.GetLogger(ctx)
+
+	// Fetch node object to set owner ref.
+	nodeObj, err := getNodeObject(ctx, info.NodeName)
+	if err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+	// Create spec for CSINodeTopology.
+	csiNodeTopologySpec := &csinodetopologyv1alpha1.CSINodeTopology{
+		ObjectMeta: metav1.ObjectMeta{
+			Name: info.NodeName,
+			OwnerReferences: []metav1.OwnerReference{
+				{
+					APIVersion: "v1",
+					Kind:       "Node",
+					Name:       nodeObj.Name,
+					UID:        nodeObj.UID,
+				},
+			},
+		},
+		Spec: csinodetopologyv1alpha1.CSINodeTopologySpec{
+			NodeID: info.NodeID,
+		},
+	}
+	// Create CSINodeTopology CR for the node.
+	err = volTopology.k8sClient.Create(ctx, csiNodeTopologySpec)
+	if err != nil {
+		if !apierrors.IsAlreadyExists(err) {
+			return nil, logger.LogNewErrorCodef(log, codes.Internal,
+				"failed to create CSINodeTopology CR. Error: %+v", err)
+		}
+	} else {
+		log.Infof("Successfully created a CSINodeTopology instance for NodeName: %q", info.NodeName)
+	}
+
+	timeoutSeconds := int64((time.Duration(getCSINodeTopologyWatchTimeoutInMin(ctx)) * time.Minute).Seconds())
```

Comment:
Why not just specify the default and max timeouts in seconds?

Reply:
We set this timeout keeping in mind that NodeGetInfo times out in 2mins according to current code. Just to keep this value in sync with the timeout in k8s code, I took the value as input in minutes. Easier for the user to change later on.